### PR TITLE
Implementing graphql based backend

### PR DESCRIPTION
**Pull request by serconde**

### CODE REVIEW

File: server/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+version: '3'
+services:
+  mongodb-bario:
+    container_name: mongodb-bario
+    image: mongo:4.2.10
```
Comment (Nasdan) [inline server/docker-compose.yml:5]: You can update to 4.4.3 version because it's the version on production server

File: server/package.json [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    "build:dev": "npm run clean && babel src -d dist --ignore=\"./src/test-runners\" --extensions \".ts\" --source-maps",
+    "type-check": "tsc --noEmit",
+    "type-check:watch": "npm run type-check -- --watch",
+    "test": "cross-env NODE_ENV=test jest -c ./config/test/jest.js --verbose",
+    "test:watch": "npm run test -- --watchAll -i --no-cache",
+    "clean": "rimraf dist"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Lemoncode/scaffolding-express-typescript.git"
```
Comment (Nasdan) [inline server/package.json:25]: this is another repository url. Check the urls below too

File: server/package.json [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  "dependencies": {
+    "apollo-server-express": "^2.19.2",
+    "dotenv": "^8.2.0",
+    "express": "^4.17.1",
+    "graphql": "^15.4.0",
+    "immer": "^8.0.1",
+    "lodash.merge": "^4.6.2",
+    "mongoose": "^5.10.11",
+    "regenerator-runtime": "^0.13.7",
+    "uuid": "^8.3.2"
```
Comment (Nasdan) [inline server/package.json:62]: You installed this library as production library, but it's used for mock repository. So we will add more size to app. Is this library actually necessary?

File: server/src/core/repository.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  getItemByPropValue: (propName: string, value: any) => IEntity;
+  setItems: (items: IEntity[]) => IEntity[];
+  deleteItem: (id: any) => IEntity[];
+}
+
+export function createArrayRepository<T extends IEntity>(
+  idGenerator: () => any,
+  initialCollection?: Array<T>
+): IArrayRepository<T> {
+  if (!!!idGenerator) throw 'An id generator must be provided';
```
Comment (Nasdan) [inline server/src/core/repository.ts:20]: Why triple exclamation mark?

File: server/src/core/repository.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import { produce } from 'immer';
+
+export interface IEntity {
```
Comment (Nasdan) [inline server/src/core/repository.ts:3]: The `I` on interfaces doesn't used in javascript. Is more a C# code style.

File: server/src/dals/bar-info/bar-info.context.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { Document, model, Schema } from 'mongoose';
+import { BarInfo } from 'dals';
+
+const barInfoSchema = new Schema({
+  _id: { type: Schema.Types.ObjectId, required: true, auto: true },
```
Comment (Nasdan) [inline server/src/dals/bar-info/bar-info.context.ts:5]: This is not necessary define this field. Mongo creates it automatically

File: server/src/dals/bar-info/bar-info.context.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  infoB: Schema.Types.String,
+  infoC: Schema.Types.String,
+  updatedAt: {
+    type: Schema.Types.Date,
+    default: Date.now
+  },  
+});
+
+// Duplicate the ID field.
+barInfoSchema.virtual('id').get(function(){
```
Comment (Nasdan) [inline server/src/dals/bar-info/bar-info.context.ts:16]: We usually delegate this logic to mappers between Model and API Models. For example:

```
const mapBarInfoFromModelToApi = (barInfo: model.BarInfo): apiModel.BarInfo => ({
  id: barInfo._id.toString(),
  infoA: barInfo.infoA,
 ....
});
```

On this way, if we have to migrate from Mongoose to Vanilla Mongo, or another driver, or another db, we can keep this mappers

File: server/src/core/repository.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+export interface IArrayRepository<IEntity> {
+  getCollection: () => IEntity[];
+  saveItem: (item: IEntity) => IEntity;
+  getItemById: (id: any) => IEntity;
+  getItemByPropValue: (propName: string, value: any) => IEntity;
+  setItems: (items: IEntity[]) => IEntity[];
+  deleteItem: (id: any) => IEntity[];
+}
+
+export function createArrayRepository<T extends IEntity>(
```
Comment (Nasdan) [inline server/src/core/repository.ts:16]: I saw below that this functionality is for mock repositories. So we will move to `common/helpers/mock-repository.helpers.ts` and rename to `createMockRepository`

File: server/src/dals/bar-info/repository/bar-info.repository.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { BarInfo, BarInfoContext } from 'dals';
+
+export const getBarInfoById = async (id: string): Promise<BarInfo> => {
+  if (!!!id) throw 'id cannot be empty';
+  return await BarInfoContext.findById(id);
```
Comment (Nasdan) [inline server/src/dals/bar-info/repository/bar-info.repository.ts:5]: We can use [lean method](https://mongoosejs.com/docs/tutorials/lean.html#using-lean) to get a plain Object instead of all mongoose instance

File: server/src/dals/bar-info/repository/bar-info.repository.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { BarInfo, BarInfoContext } from 'dals';
+
+export const getBarInfoById = async (id: string): Promise<BarInfo> => {
+  if (!!!id) throw 'id cannot be empty';
```
Comment (Nasdan) [inline server/src/dals/bar-info/repository/bar-info.repository.ts:4]: Why don't use `!id`?

File: server/src/dals/menu-category/menu-category.context.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    type: Schema.Types.Date,
+    default: Date.now,
+  },
+});
+
+const menuCategorySchema = new Schema({
+  _id: { type: Schema.Types.ObjectId, required: true, auto: true },
+  name: { type: Schema.Types.String, required: true },
+  products: [productSchema],
+  child: productSchema,
```
Comment (Nasdan) [inline server/src/dals/menu-category/menu-category.context.ts:36]: For what is this `child`?

File: server/src/dals/menu-category/menu-category.context.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    required: true,
+    index: true,
+    unique: true,
+  },
+  visible: { type: Schema.Types.Boolean, required: true },
+  description: { type: Schema.Types.String },
+  portionTypeId: { type: Schema.Types.ObjectId, required: true },
+  portions: [portionSchema],
+  child: portionSchema,
+  updatedAt: {
```
Comment (Nasdan) [inline server/src/dals/menu-category/menu-category.context.ts:26]: Why did you add this date? You don't use it on any model, isn't it?

File: server/src/pods/bar-info/bar-info.resolver.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  };
+  Mutation: {
+    saveBarInfo: GraphQLResolver<BarInfo, { barInfo: BarInfo }>;
+  };
+}
+
+export const barInfoResolver: BarInfoResolver = {
+  Query: {
+    getBarInfoById: async (parent, { id }, context): Promise<BarInfo> =>
+      await barInfoRepository.getBarInfoById(id),
```
Comment (Nasdan) [inline server/src/pods/bar-info/bar-info.resolver.ts:16]: Good place to use mappers from model to api model

File: server/src/pods/bar-info/bar-info.resolver.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  Mutation: {
+    saveBarInfo: async (parent, { barInfo }, context): Promise<BarInfo> =>
+      await barInfoRepository.saveBarInfo(barInfo),
```
Comment (Nasdan) [inline server/src/pods/bar-info/bar-info.resolver.ts:20]: Good place to use mappers from api model to model

File: server/src/common/helpers/repository.ts [hunk 15]
```diff
@@ -0,0 +1,79 @@
+import { produce } from 'immer';
```
Comment (Nasdan) [inline server/src/common/helpers/repository.ts:1]: I think `server/src/common/helpers/index.ts` is empty, isn't it?

File: server/src/common/helpers/repository.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+export interface Entity {
+  _id: any;
+}
+
+export interface IMockRepository<Entity> {
```
Comment (Nasdan) [inline server/src/common/helpers/repository.ts:7]: Remove `I` here too :)

File: server/src/dals/bar-info/repository/bar-info.mock.ts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import { createMockRepository } from 'core';
+import { BarInfo } from 'dals';
+import { v4 as uuidv4 } from 'uuid';
+
+const repository = createMockRepository<BarInfo>(() => uuidv4());
```
Comment (Nasdan) [inline server/src/dals/bar-info/repository/bar-info.mock.ts:5]: Really nice to create mock repositories :)

File: server/src/dals/bar-info/repository/bar-info.repository.ts (hunk 9 above)
Comment (Nasdan) [reply server/src/dals/bar-info/repository/bar-info.repository.ts:6]: Pending to apply here?

File: server/src/pods/bar-info/bar-info.typedefs.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    infoB: String
+    infoC: String
+  }
+
+  extend type Query {
+    getBarInfoById(id: ID!): BarInfo
+  }
+
+  input BarInfoInput {
+    id: ID
```
Comment (Nasdan) [inline server/src/pods/bar-info/bar-info.typedefs.ts:16]: The Id is required isn't it? `ID!`

File: server/src/pods/menu-categories/menu-category.typedefs.ts [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  input MenuCategoryProductInput {
+    id: ID!
+    name: String!
+    visible: Boolean!
+    portionTypeId: ID
+    portions: [PortionInput]
+  } 
+
+  input MenuCategoryInput {
+    id: ID
```
Comment (Nasdan) [inline server/src/pods/menu-categories/menu-category.typedefs.ts:54]: Is id required?

File: server/src/pods/menu-categories/menu-category.typedefs.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    getProductById(id: ID!): Product
+  }
+
+  input PortionInput {
+    id: ID!
+    price: Float!
+  }
+
+  input ProductInput {
+    id: ID
```
Comment (Nasdan) [inline server/src/pods/menu-categories/menu-category.typedefs.ts:37]: is id required?

File: server/src/dals/menu-category/repository/index.ts [hunk 21]
```diff
@@ -0,0 +1,8 @@
+import { envConstants } from 'core';
+import * as mockRepository from './menu-category.mock';
+import * as repository from './menu-category.repository';
+import { MenuCategoryRepositoryContract } from './menu-category.contract';
+
+export const MenuCategoryRepository: MenuCategoryRepositoryContract = envConstants.isMockRepository
```
Comment (Nasdan) [inline server/src/dals/menu-category/repository/index.ts:6]: `menuCategoryRepository`

File: server/src/pods/bar-info/bar-info.typedefs.ts (hunk 18 above)
Comment (serconde) [reply server/src/pods/bar-info/bar-info.typedefs.ts:16]: It isn't required as this method is used to upsert, so when the ID is empty, a new one is created when the item is saved

File: server/src/pods/menu-categories/menu-category.typedefs.ts (hunk 20 above)
Comment (serconde) [reply server/src/pods/menu-categories/menu-category.typedefs.ts:37]: Same as above

File: server/src/pods/menu-categories/menu-category.typedefs.ts (hunk 19 above)
Comment (serconde) [reply server/src/pods/menu-categories/menu-category.typedefs.ts:54]: Used for upserts so it isn't necessary
